src/Database/LSMTree/Internal/RawBytes.hs

```diff
@@ -172,14 +175,29 @@ drop = coerce VP.drop
 --
 -- The /top/ corresponds to the most significant bit (big-endian).
 --
--- PRECONDITION: The byte-size of the raw bytes should be at least 8 bytes.
+-- If the number of bits is smaller than @64@, then any missing bits default to
+-- @0@s.
+--
+-- >>> showHex (topBits64 (pack [1,0,0,0,0,0,0,0])) ""
+-- "100000000000000"
+--
+-- >>> showHex (topBits64 (pack [1,0,0])) ""
+-- "100000000000000"
 --
 -- TODO: optimisation ideas: use unsafe shift/byteswap primops, look at GHC
 -- core, find other opportunities for using primops.
 --
 topBits64 :: RawBytes -> Word64
-topBits64 rb@(RawBytes (VP.Vector (I# off#) _size (ByteArray k#))) =
-    assert (size rb >= 8) $ toWord64 (indexWord8ArrayAsWord64# k# off#)
+topBits64 rb@(RawBytes v@(VP.Vector (I# off#) _size (ByteArray k#)))
+  | n >= 8
+  = toWord64 (indexWord8ArrayAsWord64# k# off#)
+  | otherwise
+  = VP.foldl' f 0 v `unsafeShiftL` ((8 - n) * 8)
+  where
+    !n = size rb
+
+    f :: Word64 -> Word8 -> Word64
+    f acc w = acc `unsafeShiftL` 8 + fromIntegral w
```

Comment:
I think this can be done more straightforwardly and efficiently, without a loop

You'd need to:
1. use `indexWord8ArrayAsWord64#` to do an unaligned read
2. use the `size rb` to generate a bitmask, to zero out all the undefined bits: ((8 - n) * 8) bits of trailing 0s
3. apply the bitmask with `.&.`

I don't think this will be backwards, but if so, use a `byteSwap64` between step 1 and 2

Reply:
I tried not to read outside of the array for fear of segfaults, but I could try your approach